FlowCal/gate.py

```diff
@@ -334,17 +364,35 @@ def density2d(data, channels=[0,1],
     if data_ch.shape[0] <= 1:
         raise ValueError('data should have more than one event')
 
-    # Extract default bins if necessary
-    if (bins is None and hasattr(data_ch, 'hist_bin_edges')
-            and data_ch.hist_bin_edges(0) is not None
-            and data_ch.hist_bin_edges(1) is not None):
-        bins = np.array(data_ch.hist_bin_edges())
+    # If ``data_ch.hist_bins()`` exists, obtain bin edges from it if
+    # necessary.
+    if hasattr(data_ch, 'hist_bins'):
+        # Check whether `bins` contains information for one or two axes
+        if hasattr(bins, '__iter__') and len(bins)==2:
+            # `bins` contains separate information for both axes
+            # If bins for the X axis is not an iterable, get bin edges from
+            # ``data_ch.hist_bins()``.
+            if not hasattr(bins[0], '__iter__'):
+                bins[0] = data_ch.hist_bins(channels=0,
+                                              nbins=bins[0],
+                                              log=xlog)
+            # If bins for the Y axis is not an iterable, get bin edges from
+            # ``data_ch.hist_bins()``.
+            if not hasattr(bins[1], '__iter__'):
+                bins[1] = data_ch.hist_bins(channels=1,
+                                              nbins=bins[1],
+                                              log=ylog)
+        else:
+            # `bins` contains information for one axis, which will be used
+            # twice.
+            # If bins is not an iterable, get bin edges from
+            # ``data_ch.hist_bins()``.
+            if not hasattr(bins, '__iter__'):
+                bins = [data_ch.hist_bins(channels=0, nbins=bins, log=xlog),
+                        data_ch.hist_bins(channels=1, nbins=bins, log=ylog)]
 
     # Make 2D histogram
-    if bins is not None:
-        H,xe,ye = np.histogram2d(data_ch[:,0], data_ch[:,1], bins=bins)
-    else:
-        H,xe,ye = np.histogram2d(data_ch[:,0], data_ch[:,1])
```

Comment:
this `np.histogram2d` call will now fail of `bins` is `None`. Can happen if you pass in a numpy array (instead of an `FCSData` object) and don't specify `bins`.

spoke with @castillohair in person, but not sure if we want to update `bins` default for the `plot` and `gate` density functions or add the `if-else` statement back in.